[awlarpi] iP

awlarpi opened the pull request:
# Bob

> “Your mind is for having ideas, not holding them.” – David Allen ([source](https://gettingthingsdone.com/))

Bob frees your mind of having to remember things you need to do. It's:

- text-based
- easy to learn
- **FAST** and *SUPER FAST* to use

All you need to do is:

1. [Download it from here](https://github.com/nus-cs2103-AY2425S2/ip/pull/371).
2. Double-click it.
3. Add your tasks.
4. Let it manage your tasks for you 😉.

And it is FREE!

## Features:

- [x] Managing tasks
- [ ] Managing deadlines (coming soon)
- [ ] Reminders (coming soon)

If you are a Java programmer, you can use it to practice Java too. Here's the main method:

```java
public class Main {
    public static void main(String[] args) {
        Application.launch(MainApp.class, args);
    }
}

Commit:
Add Gradle support

Eclipse-Dominator committed:
Bump gradle and lib version

aureliony committed:
build.gradle: Prevent generating a second JAR file

In build.gradle, the dependencies on distZip and/or distTar causes
the shadowJar task to generate a second JAR file for which the
mainClass.set("seedu.duke.Duke") does not take effect.
Hence, this additional JAR file cannot be run.
For this product, there is no need to generate a second JAR file
to begin with.

Let's remove this dependency from the build.gradle to prevent the
shadowJar task from generating the extra JAR file.

awlarpi committed:
Rename Duke to Bob, and add greet functionality

awlarpi committed:
Add echo functionality to Duke

awlarpi committed:
Add ability to store and display user input

awlarpi committed:
Refactor Bob to use Task class for task management and add mark/unmark functionality

awlarpi committed:
Add task management features: implement ToDos, Deadlines, and Events classes

Add three new classes to manage different types of tasks: ToDos, Deadlines, and Events.
These classes extend the Task class and provide additional functionality to manage tasks.

- ToDos: Represents a simple task without any date/time attached to it.
- Deadlines: Represents a task that needs to be done before a specific date/time.
- Events: Represents a task that starts at a specific time and ends at a specific time.

awlarpi committed:
Update test input and expected output

awlarpi committed:
Add error handling with user feedback

awlarpi committed:
Add delete functionality for tasks and update expected outputs

awlarpi committed:
Add CommandType enum for task management commands

awlarpi committed:
Update AI.md to include observations on Github Copilot usage

awlarpi committed:
Created using Colab

awlarpi committed:
Add local data persistence

awlarpi committed:
Add run instructions to README.md

awlarpi committed:
Revert "Created using Colab"

This reverts commit 5d9394ec53b55aef849f1cb05cfac4317dca37e0.

awlarpi committed:
Merge branch 'master' into branch-Level-7

awlarpi committed:
Add date handling to Deadlines and Events

awlarpi committed:
Extract Ui, Storage, Parser and TaskList classes

awlarpi committed:
Place all java files into the Bob package

awlarpi committed:
Merge remote-tracking branch 'remotes/origin/add-gradle-support'

awlarpi committed:
Update build.gradle to reflect Bob package

awlarpi committed:
Add  junit tests

awlarpi committed:
Refactor into different packages and extract Command as an interface

awlarpi committed:
Add JavaDoc comments to command classes and models for better documentation

awlarpi committed:
Add formatting standard

awlarpi committed:
Add FindCommand and implement task searching functionality

awlarpi committed:
Merge branch 'branch-A-CodingStandard'

awlarpi committed:
Merge branch 'branch-Level-9'

awlarpi committed:
Refactor Storage class to simplify task initialization and error handling

awlarpi committed:
Add JavaDoc for findTasksByKeyword

zhenglong1603 commented on gradle/wrapper/gradle-wrapper.properties:
```diff
@@ -0,0 +1,7 @@
+distributionBase=GRADLE_USER_HOME
```
maybe consider adding the whole gradle to the gitignore as well?

awlarpi replied:
Yes good suggestion.

zhenglong1603 commented on src/main/java/bob/Bob.java:
```diff
@@ -0,0 +1,68 @@
+package bob;
```
consider adding some of your own exceptions as well!

awlarpi replied:
Yes I will try making my own Exceptions, thanks for the suggestion.

zhenglong1603 commented on src/main/java/bob/models/Task.java:
```diff
@@ -0,0 +1,59 @@
+/**
+ * Represents a task.
+ */
+public class Task implements Serializable {
```
consider making this an abstract class! you've already got the three task types, there's probably no reason for a bare Task object to be instantiated at this point.

awlarpi replied:
Yes good idea.

zhenglong1603 commented on src/main/java/bob/models/TaskList.java:
```diff
@@ -0,0 +1,103 @@
+     *
+     * @return The list of tasks.
+     */
+    public ArrayList<Task> getTasks() {
```
i'm not sure if this is properly preserving encapsulation? Maybe consider returning a copy instead?

awlarpi replied:
Yes that's the ideal case, I will consider.

zhenglong1603 reviewed (commented):
Overall very gd job ! It was a pleasure reading this and keep up the good work:)

chenxu20 commented on src/main/java/bob/commands/Parser.java:
```diff
@@ -0,0 +1,57 @@
+            }
+            return new DeadlineCommand(deadlineParts[0].trim(), deadlineParts[1].trim());
+        case "event":
+            String[] eventParts = arguments.split(" /from ");
```
Perhaps can specify the exact event parts, like descriptionAndTime

chenxu20 commented on src/main/java/bob/models/Task.java:
```diff
@@ -0,0 +1,59 @@
+ * Represents a task.
+ */
+public class Task implements Serializable {
+    private static final long serialVersionUID = 1L;
```
Since this variable has a final modifier, I think you can consider capitalising the entire name.

chenxu20 commented on src/main/java/bob/models/Deadline.java:
```diff
@@ -0,0 +1,41 @@
+ */
+public class Deadline extends Task {
+    private LocalDateTime by;
+    private static final DateTimeFormatter INPUT_FORMATTER =
```
Nice job for making the name of your constant all upper case.

awlarpi replied:
Thank you!

chenxu20 reviewed (commented):
LGTM. Most of your names are clear and concise.

awlarpi committed:
Add Checkstyle and Fix Style Issues

- Integrated Checkstyle into the project to enforce coding standards.
- Fixed various style issues across the codebase to comply with Checkstyle rules.
- Updated the build configuration to include Checkstyle checks.
- Ensured all existing code passes the Checkstyle validation.

awlarpi committed:
Add GUI for Bob using JavaFX

- Implemented the main application class `Main` to launch the JavaFX application.
- Created `MainWindow` controller to handle user interactions and display responses.
- Added FXML files `MainWindow.fxml` and `DialogBox.fxml` for the layout of the main window and dialog boxes.
- Updated `Bob` class to integrate with the GUI and handle user input.
- Included necessary JavaFX dependencies in `build.gradle`.
- Ensured the application exits gracefully with `ByeCommand`.

awlarpi committed:
Add varargs method to add multiple tasks to TaskList

- Implemented addTasks(Task... tasks) method to allow adding multiple tasks at once.
- Updated TaskList class to include the new method.
- This change provides flexibility in adding tasks and simplifies the code when dealing with multiple tasks.

awlarpi committed:
Add assertion to ensure command is not null in getResponse method

- Added an assertion to check that the command parsed from the input is not null.
- This helps to catch potential issues early and provides a clear error message for debugging.
- Ensures that the command object is valid before attempting to execute it.

awlarpi committed:
Refactor code to improve code quality

- Extracted overly long method in Parser to improve readability and maintainability.
- Simplified TaskList toString method for better clarity.
- Refactored Bob class to handle exceptions more gracefully and improve logging.

These changes enhance the overall code quality by making the codebase easier to understand and maintain.

awlarpi committed:
Refactor TaskList methods to use Java Streams

- Updated `getCompletedTasks` method to use streams for filtering completed tasks.
- Updated `getTaskDescriptions` method to use streams for mapping task descriptions.
- Improved code readability and conciseness by leveraging Java Streams API.

awlarpi committed:
Merge pull request #4 from awlarpi/branch-A-Streams

Refactor TaskList methods to use Java Streams

awlarpi committed:
Merge branch 'master' into branch-A-CodeQuality
- Merge branch master branch to keep it up to date.

awlarpi committed:
Merge pull request #3 from awlarpi/branch-A-CodeQuality

Refactor code to improve code quality

awlarpi committed:
Merge branch 'master' into branch-A-Assertions

awlarpi committed:
Merge pull request #2 from awlarpi/branch-A-Assertions

Add assertions

awlarpi committed:
Add shortcut commands for todo, deadline, and event

- Added support for shortcut commands: "t" for "todo",
  "d" for "deadline", and "e" for "event".
- Updated the parse method to handle these shortcuts.

awlarpi committed:
Add attribution for code adaptations from SE Education tutorial

- Added attribution comment for code adapted from
  SE Education tutorial
- https://se-education.org/guides/tutorials/javaFxPart4.html

awlarpi committed:
Create custom CommandException class and subclasses

- Added CommandException as a base class for command-related errors.
- Created specific subclasses like TaskNumberOutOfBoundsException.
- Updated code to use these custom exceptions for better error handling.

awlarpi committed:
Add user guide

- Added instructions for adding deadlines, events, and todos
- Included examples and expected outputs for each command
- Detailed steps for marking, unmarking, and deleting tasks
- Provided instructions for listing and finding tasks
- Added section on exiting the application
- Included steps to run the application via JAR file or terminal
- Listed prerequisites for running the application

awlarpi committed:
Update README.md to include shortcut commands

- Added shortcut commands for adding deadlines, events, and todos
- Updated examples to include both full and shortcut commands
- Ensured expected output examples reflect the use of shortcuts

awlarpi committed:
Trim leading and trailing whitespace from user input

- Updated `parse` method in `Parser.java` to trim leading and trailing
  whitespace from user input.
- Ensures commands are parsed correctly even with extra spaces at the
  beginning or end of input.
- Added `input = input.trim();` at the start of the `parse` method.
- Preserves intentional whitespace within command arguments.

This update improves command parsing by handling extra whitespace.
